In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

En pimera instancia se realizo una exploracio0n de la codificacion del archivo, dado que saltaba error en la codificacion del mismo, e identiificamos que el archivo de encuentra codificado en ISO-8859-1

In [2]:
try:
    df = pd.read_csv("Shark_attacks.csv", encoding="utf-8")
    print("El archivo se leyó con codificación UTF-8.")
except UnicodeDecodeError:
    print("La codificación no es UTF-8. Intentando con ISO-8859-1.")
    try:
        df = pd.read_csv("Shark_attacks.csv", encoding="ISO-8859-1")
        print("El archivo se leyó con codificación ISO-8859-1.")
    except UnicodeDecodeError:
        print("No se pudo leer el archivo. Revisa la codificación.")


La codificación no es UTF-8. Intentando con ISO-8859-1.
El archivo se leyó con codificación ISO-8859-1.


In [3]:
df = pd.read_csv('Shark_attacks.csv', encoding='ISO-8859-1', low_memory=False)



averiguamos el numero total de registros y columnas, generando la dimension total del dataset

In [4]:
print(df.shape)

(25614, 22)


exploramos los registros nulos

In [5]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25614 entries, 0 to 25613
Data columns (total 22 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             6095 non-null   object 
 1   Date                    6094 non-null   object 
 2   Year                    6092 non-null   float64
 3   Type                    6090 non-null   object 
 4   Country                 6048 non-null   object 
 5   Area                    5682 non-null   object 
 6   Location                5583 non-null   object 
 7   Activity                5559 non-null   object 
 8   Name                    5888 non-null   object 
 9   Sex                     5517 non-null   object 
 10  Age                     3374 non-null   object 
 11  Injury                  6066 non-null   object 
 12  Fatal (Y/N)             6064 non-null   object 
 13  Time                    2848 non-null   object 
 14  Species                 3100 non-null 

In [6]:
df.head()

Case Number                  Date    Year        Type         Country  \
0    2017.06.11            2017-06-11  2017.0  Unprovoked       AUSTRALIA   
1  2017.06.10.b            2017-06-10  2017.0  Unprovoked       AUSTRALIA   
2  2017.06.10.a            2017-06-10  2017.0  Unprovoked             USA   
3  2017.06.07.R  Reported 07-Jun-2017  2017.0  Unprovoked  UNITED KINGDOM   
4    2017.06.04            2017-06-04  2017.0  Unprovoked             USA   

                Area                                         Location  \
0  Western Australia                         Point Casuarina, Bunbury   
1           Victoria                    Flinders, Mornington Penisula   
2            Florida                      Ponce Inlet, Volusia County   
3        South Devon                                    Bantham Beach   
4            Florida  Middle Sambo Reef off Boca Chica, Monroe County   

        Activity            Name Sex   ... Fatal (Y/N)   Time  \
0  Body boarding       Paul Goff    M  ...           N  08h30   
1        Surfing          female    F  ...           N  15h45   
2        Surfing     Bryan Brock    M  ...           N  10h00   
3       Surfing     Rich Thomson    M  ...           N    NaN   
4   Spearfishing  Parker Simpson    M  ...           N    NaN   

                            Species                  Investigator or Source  \
0                   White shark, 4 m                    WA Today, 6/11/2017   
1                       7 gill shark                                    NaN   
2                                NaN  Daytona Beach News-Journal, 6/10/2017   
3  3m shark, probably a smooth hound                         C. Moore, GSAF   
4                           8' shark                    Nine News, 6/7/2017   

                         pdf  \
0        2017.06.11-Goff.pdf   
1  2017.06.10.b-Flinders.pdf   
2     2017.06.10.a-Brock.pdf   
3   2017.06.07.R-Thomson.pdf   
4     2017.06.04-Simpson.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2017.06.11   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  2017.06.10.b   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  2017.06.10.a   
3  http://sharkattackfile.net/spreadsheets/pdf_di...  2017.06.07.R   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2017.06.04   

  Case Number.2 original order  
0    2017.06.11         6095.0  
1  2017.06.10.b         6094.0  
2  2017.06.10.a         6093.0  
3  2017.06.07.R         6092.0  
4    2017.06.04         6091.0  

[5 rows x 22 columns]

Buscamos Valores unicos en la columna Case Number para saber si cada caso es distinto

In [7]:
valores_unicos = df["Case Number"].unique()
print(len(valores_unicos))

6080


Como tenemos un total de 6095 registros y 6080 valores son unicos, damos por supuesto que todos los casos son distintos 

despues de explorar la tabla, y observar la gran cantidad de valores unicos por columna, decidimos realizar una limpieza de las columnas que no serian utiles en nuestro analisis, quedando unicamente las columnas siguientes:

In [8]:
dflimpio = df[["Year","Type","Country","Sex ","Age","Fatal (Y/N)","Date","Activity","Species "]]

In [9]:
dflimpio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25614 entries, 0 to 25613
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         6092 non-null   float64
 1   Type         6090 non-null   object 
 2   Country      6048 non-null   object 
 3   Sex          5517 non-null   object 
 4   Age          3374 non-null   object 
 5   Fatal (Y/N)  6064 non-null   object 
 6   Date         6094 non-null   object 
 7   Activity     5559 non-null   object 
 8   Species      3100 non-null   object 
dtypes: float64(1), object(8)
memory usage: 1.8+ MB


agrupamos las diferentes actividades en 8 grupos definidos y asignamos Unknown a los valores desconocidos

In [10]:



# Define a dictionary to map each specific activity to broader categories
activity_groups = {
    'Surfing': 'Water Activity',
    'Swimming': 'Water Activity',
    'Snorkeling': 'Water Activity',
    'Diving': 'Water Activity',
    'Fishing': 'Fishing',
    'Spearfishing': 'Fishing',
    'Boating': 'Boating',
    'Kayaking': 'Boating',
    'Standing': 'Other',  # Adjust according to each unique value in `unique_values`
    # Add more activities as needed based on the unique values
}

# Map the activities to the broader categories and create a new "Category" column
dflimpio['Categoryactivity'] = dflimpio['Activity'].map(activity_groups)

# Fill any NaN values in the "Category" column with "Unknown" or "Other" for unmapped activities
dflimpio['Categoryactivity'] = dflimpio['Categoryactivity'].fillna("Unknown")




print(dflimpio.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25614 entries, 0 to 25613
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Year              6092 non-null   float64
 1   Type              6090 non-null   object 
 2   Country           6048 non-null   object 
 3   Sex               5517 non-null   object 
 4   Age               3374 non-null   object 
 5   Fatal (Y/N)       6064 non-null   object 
 6   Date              6094 non-null   object 
 7   Activity          5559 non-null   object 
 8   Species           3100 non-null   object 
 9   Categoryactivity  25614 non-null  object 
dtypes: float64(1), object(9)
memory usage: 2.0+ MB
None


/tmp/ipykernel_45368/520225145.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dflimpio['Categoryactivity'] = dflimpio['Activity'].map(activity_groups)
/tmp/ipykernel_45368/520225145.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dflimpio['Categoryactivity'] = dflimpio['Categoryactivity'].fillna("Unknown")


eliminamos la columna antigua activity 

In [11]:
dflimpio = dflimpio.drop("Activity", axis=1)

dflimpio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25614 entries, 0 to 25613
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Year              6092 non-null   float64
 1   Type              6090 non-null   object 
 2   Country           6048 non-null   object 
 3   Sex               5517 non-null   object 
 4   Age               3374 non-null   object 
 5   Fatal (Y/N)       6064 non-null   object 
 6   Date              6094 non-null   object 
 7   Species           3100 non-null   object 
 8   Categoryactivity  25614 non-null  object 
dtypes: float64(1), object(8)
memory usage: 1.8+ MB


agrupamos las diferentes especies en 16 grupos definidos y asignamos Unknown a los valores desconocidos

In [12]:
import re
def categorize_shark(description):
    # Define common shark types and associated keywords
    keywords = {
        'white': ['WHITE', 'GREAT WHITE'],
        'tiger': ['TIGER'],
        'hammerhead': ['HAMMERHEAD'],
        'thresher': ['THRESHER'],
        'mako': ['MAKO'],
        'bull': ['BULL'],
        'lemon': ['LEMON'],
        'whitetip': ['WHITE TIP', 'WHITETIP'],
        'blacktip': ['BLACK TIP', 'BLACKTIP'],
        'blue': ['BLUE'],
        'grey': ['GREY', 'GRAY NURSE', 'GREY NURSE'],
        'nurse': ['NURSE'],
        'thresher': ['THRESHER'],
        'sevengill': ['7 GILL', 'SEVEN GILL'],
        'porbeagle': ['PORBEAGLE'],
        'hammer': ['HAMMERHEAD']
    }

    # Standardize the text to uppercase for matching
    description = description.upper()

    # Check for each keyword in the description
    for category, terms in keywords.items():
        if any(re.search(term, description) for term in terms):
            return category

    # If no keywords match, label as "unknown"
    return 'unknown'


dflimpio['categoryspecie'] = dflimpio['Species '].fillna('').apply(categorize_shark)


dflimpio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25614 entries, 0 to 25613
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Year              6092 non-null   float64
 1   Type              6090 non-null   object 
 2   Country           6048 non-null   object 
 3   Sex               5517 non-null   object 
 4   Age               3374 non-null   object 
 5   Fatal (Y/N)       6064 non-null   object 
 6   Date              6094 non-null   object 
 7   Species           3100 non-null   object 
 8   Categoryactivity  25614 non-null  object 
 9   categoryspecie    25614 non-null  object 
dtypes: float64(1), object(9)
memory usage: 2.0+ MB


eliminamos las columnas que ya no neceitamos

In [13]:
dflimpio = dflimpio.drop("Species ", axis=1)
dflimpio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25614 entries, 0 to 25613
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Year              6092 non-null   float64
 1   Type              6090 non-null   object 
 2   Country           6048 non-null   object 
 3   Sex               5517 non-null   object 
 4   Age               3374 non-null   object 
 5   Fatal (Y/N)       6064 non-null   object 
 6   Date              6094 non-null   object 
 7   Categoryactivity  25614 non-null  object 
 8   categoryspecie    25614 non-null  object 
dtypes: float64(1), object(8)
memory usage: 1.8+ MB


renombramos las columnas que creamos para que mantengan coherencia

In [14]:
dflimpio["Activity"] = dflimpio["Categoryactivity"]


In [15]:
dflimpio = dflimpio.drop("Categoryactivity", axis=1)

In [16]:
dflimpio["Specie"] = dflimpio["categoryspecie"]

In [17]:
dflimpio = dflimpio.drop("categoryspecie", axis=1)

In [18]:
dflimpio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25614 entries, 0 to 25613
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         6092 non-null   float64
 1   Type         6090 non-null   object 
 2   Country      6048 non-null   object 
 3   Sex          5517 non-null   object 
 4   Age          3374 non-null   object 
 5   Fatal (Y/N)  6064 non-null   object 
 6   Date         6094 non-null   object 
 7   Activity     25614 non-null  object 
 8   Specie       25614 non-null  object 
dtypes: float64(1), object(8)
memory usage: 1.8+ MB


Para extrae el mes de la columna Date,y almacenarlo en una nueva columna llamada Month hacemos lo siguiente

In [19]:
dflimpio["Date"] = dflimpio["Date"].str.upper().str.strip()

In [20]:

meses_dict = {
    'JAN': 1.0,
    'FEB': 2.0,
    'MAR': 3.0,
    'APR': 4.0,
    'MAY': 5.0,
    'JUN': 6.0,
    'JUL': 7.0,
    'AUG': 8.0,
    'SEP': 9.0,
    'OCT': 10.0,
    'NOV': 11.0,
    'DEC': 12.0
}

def extract_month(date):
    if not isinstance(date, str) or len(date) < 4:  # Asegurarse de que sea una cadena y tenga longitud mínima
        return None

    parsed_date = pd.to_datetime(date, errors='coerce')
    if pd.notnull(parsed_date):
        return parsed_date.month

    partes = date.split('-')
    if len(partes) > 1:  # Verificar que haya al menos dos partes
        mes_abreviatura = partes[1][0:3]  # TOMAR LOS TRES PRIMEROS CARACTERES DEL MES
        # Buscar el mes en el diccionario y devolver el número correspondiente
        return meses_dict.get(mes_abreviatura, None)

    return None  # Devolver None si no se cumple ninguna condición

dflimpio['Month'] = dflimpio['Date'].apply(extract_month)  # Cambia 'Date' al nombre real de la columna de fechas




In [21]:
dflimpio.info()
dflimpio.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25614 entries, 0 to 25613
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         6092 non-null   float64
 1   Type         6090 non-null   object 
 2   Country      6048 non-null   object 
 3   Sex          5517 non-null   object 
 4   Age          3374 non-null   object 
 5   Fatal (Y/N)  6064 non-null   object 
 6   Date         6094 non-null   object 
 7   Activity     25614 non-null  object 
 8   Specie       25614 non-null  object 
 9   Month        5825 non-null   float64
dtypes: float64(2), object(8)
memory usage: 2.0+ MB


Year        Type         Country Sex   Age Fatal (Y/N)  \
0  2017.0  Unprovoked       AUSTRALIA    M   48           N   
1  2017.0  Unprovoked       AUSTRALIA    F  NaN           N   
2  2017.0  Unprovoked             USA    M   19           N   
3  2017.0  Unprovoked  UNITED KINGDOM    M   30           N   
4  2017.0  Unprovoked             USA    M  NaN           N   

                   Date        Activity     Specie  Month  
0            2017-06-11         Unknown      white    6.0  
1            2017-06-10  Water Activity  sevengill    6.0  
2            2017-06-10  Water Activity    unknown    6.0  
3  REPORTED 07-JUN-2017         Unknown    unknown    6.0  
4            2017-06-04         Fishing    unknown    6.0

eliminamos los datos nulos

In [22]:
df_clear = dflimpio.dropna()

In [23]:
df_clear.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3281 entries, 0 to 5953
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         3281 non-null   float64
 1   Type         3281 non-null   object 
 2   Country      3281 non-null   object 
 3   Sex          3281 non-null   object 
 4   Age          3281 non-null   object 
 5   Fatal (Y/N)  3281 non-null   object 
 6   Date         3281 non-null   object 
 7   Activity     3281 non-null   object 
 8   Specie       3281 non-null   object 
 9   Month        3281 non-null   float64
dtypes: float64(2), object(8)
memory usage: 282.0+ KB


In [24]:
df_clear.head()

Year        Type         Country Sex  Age Fatal (Y/N)  \
0  2017.0  Unprovoked       AUSTRALIA    M  48           N   
2  2017.0  Unprovoked             USA    M  19           N   
3  2017.0  Unprovoked  UNITED KINGDOM    M  30           N   
5  2017.0  Unprovoked         BAHAMAS    F  32           N   
6  2017.0    Provoked             USA    F  20           N   

                   Date        Activity   Specie  Month  
0            2017-06-11         Unknown    white    6.0  
2            2017-06-10  Water Activity  unknown    6.0  
3  REPORTED 07-JUN-2017         Unknown  unknown    6.0  
5            2017-06-02  Water Activity    tiger    6.0  
6            2017-05-30         Unknown  unknown    5.0

Vamos a limpiar la columna de la edad.

Convertimos la columna a numerico y los valores no numericos se transforman en NaN

In [25]:

df_clear['Age'] = pd.to_numeric(df_clear['Age'], errors='coerce')

print(df_clear['Age'])
nulos = df_clear["Age"].isna().sum()
print(nulos)

0       48.0
2       19.0
3       30.0
5       32.0
6       20.0
        ... 
5900    17.0
5917    22.0
5937     NaN
5946    14.0
5953    19.0
Name: Age, Length: 3281, dtype: float64
90


/tmp/ipykernel_45368/3932975453.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clear['Age'] = pd.to_numeric(df_clear['Age'], errors='coerce')


Dado que son pocos registros los que resultan nulos las edades, comparados con el resto de datos. 

Eliminamos los valores NaN en la columna Age

In [26]:
df_clear = df_clear.dropna(subset=['Age'])

Convertimos el tipo de Dato de la columna Age de float a Int

In [27]:
df_clear['Age'] = df_clear['Age'].astype(int)


In [28]:
df_clear.head()
df_clear.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3191 entries, 0 to 5953
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         3191 non-null   float64
 1   Type         3191 non-null   object 
 2   Country      3191 non-null   object 
 3   Sex          3191 non-null   object 
 4   Age          3191 non-null   int64  
 5   Fatal (Y/N)  3191 non-null   object 
 6   Date         3191 non-null   object 
 7   Activity     3191 non-null   object 
 8   Specie       3191 non-null   object 
 9   Month        3191 non-null   float64
dtypes: float64(2), int64(1), object(7)
memory usage: 274.2+ KB


Vamos a limpiar la columna de los años.

Miramos los valores unicos 

In [29]:
valores_unicos = df["Year"].unique()
print(valores_unicos)

[2017.   nan 2016. 2015. 2014. 2013. 2012. 2011. 2010. 2009. 2008. 2007.
 2006. 2005. 2004. 2003. 2002. 2001. 2000. 1999. 1998. 1997. 1996. 1995.
 1984. 1994. 1993. 1992. 1991. 1990. 1989. 1969. 1988. 1987. 1986. 1985.
 1983. 1982. 1981. 1980. 1979. 1978. 1977. 1976. 1975. 1974. 1973. 1972.
 1971. 1970. 1968. 1967. 1966. 1965. 1964. 1963. 1962. 1961. 1960. 1959.
 1958. 1957. 1956. 1955. 1954. 1953. 1952. 1951. 1950. 1949. 1948. 1848.
 1947. 1946. 1945. 1944. 1943. 1942. 1941. 1940. 1939. 1938. 1937. 1936.
 1935. 1934. 1933. 1932. 1931. 1930. 1929. 1928. 1927. 1926. 1925. 1924.
 1923. 1922. 1921. 1920. 1919. 1918. 1917. 1916. 1915. 1914. 1913. 1912.
 1911. 1910. 1909. 1908. 1907. 1906. 1905. 1904. 1903. 1902. 1901. 1900.
 1899. 1898. 1897. 1896. 1895. 1894. 1893. 1892. 1891. 1890. 1889. 1888.
 1887. 1886. 1885. 1884. 1883. 1882. 1881. 1880. 1879. 1878. 1877. 1876.
 1875. 1874. 1873. 1872. 1871. 1870. 1869. 1868. 1867. 1866. 1865. 1864.
 1863. 1862. 1861. 1860. 1859. 1858. 1856. 1855. 18

Limpiamos los registros con fechas NaN y con fechas anteriores o iguales a 500

In [30]:
df_clear = df_clear[df_clear["Year"]>500]


Ahora transformamos el tipo de dato de la columna de float a int 

In [31]:
df_clear['Year'] = df_clear['Year'].astype(int)

convertimos el dato int a datetype 

In [32]:
df_clear['Year'] = pd.to_datetime(df_clear['Year'], format='%Y', errors='coerce')
df_clear['Year'] = df_clear['Year'].dt.year

como se puede observar, despues de trabajar con la columna Year el tamaño del dataset se ha reducido de 3221 a 3209

In [33]:
df_clear.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3191 entries, 0 to 5953
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         3191 non-null   int32  
 1   Type         3191 non-null   object 
 2   Country      3191 non-null   object 
 3   Sex          3191 non-null   object 
 4   Age          3191 non-null   int64  
 5   Fatal (Y/N)  3191 non-null   object 
 6   Date         3191 non-null   object 
 7   Activity     3191 non-null   object 
 8   Specie       3191 non-null   object 
 9   Month        3191 non-null   float64
dtypes: float64(1), int32(1), int64(1), object(7)
memory usage: 261.8+ KB


Ahora trabajamos con la columns Type.

Sacamos los valores unicos para analizarlos

In [34]:
valores_unicos = df_clear["Type"].unique()
print(valores_unicos)

['Unprovoked' 'Provoked' 'Invalid' 'Boat' 'Sea Disaster' 'Boating']


en esta columna solo se realizo la operacion de cambio de tipo a categoria

In [35]:
df_clear['Type'] = df_clear['Type'].astype('category')

df_clear.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3191 entries, 0 to 5953
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Year         3191 non-null   int32   
 1   Type         3191 non-null   category
 2   Country      3191 non-null   object  
 3   Sex          3191 non-null   object  
 4   Age          3191 non-null   int64   
 5   Fatal (Y/N)  3191 non-null   object  
 6   Date         3191 non-null   object  
 7   Activity     3191 non-null   object  
 8   Specie       3191 non-null   object  
 9   Month        3191 non-null   float64 
dtypes: category(1), float64(1), int32(1), int64(1), object(6)
memory usage: 240.2+ KB


Ahora pasamos a la columna de Country

In [36]:
valores_unicos = df_clear["Country"].unique()
print(valores_unicos)

['AUSTRALIA' 'USA' 'UNITED KINGDOM' 'BAHAMAS' 'UNITED ARAB EMIRATES'
 'MEXICO' 'SOUTH AFRICA' 'REUNION' 'ST HELENA, British overseas territory'
 'INDONESIA' 'England' 'NEW CALEDONIA' 'SPAIN' 'CHINA' 'JAPAN' 'COLUMBIA'
 'EGYPT' 'NEW ZEALAND' 'FRENCH POLYNESIA' 'CAPE VERDE' 'Fiji' 'BRAZIL'
 'CAYMAN ISLANDS' 'ARUBA' 'ECUADOR' 'MOZAMBIQUE' 'THAILAND' 'PUERTO RICO'
 'ITALY' 'MAURITIUS' 'ST. MARTIN' 'PAPUA NEW GUINEA' 'KIRIBATI' 'ISRAEL'
 'FRANCE' 'DIEGO GARCIA' 'JAMAICA' 'PALESTINIAN TERRITORIES' 'GUAM'
 'SEYCHELLES' 'NIGERIA' 'TONGA' 'SCOTLAND' 'TRINIDAD & TOBAGO' 'CANADA'
 'FIJI' 'SAUDI ARABIA' 'TAIWAN' 'DOMINICAN REPUBLIC' 'KENYA' 'RUSSIA'
 'PHILIPPINES' 'TURKS & CAICOS' 'COSTA RICA' 'UNITED ARAB EMIRATES (UAE)'
 'SAMOA' 'SOLOMON ISLANDS' 'SOUTH KOREA' 'MALTA' 'VIETNAM' 'MADAGASCAR'
 'PANAMA' 'AZORES' 'CUBA' 'CROATIA' 'ENGLAND' 'BRITISH VIRGIN ISLANDS'
 'NORWAY' 'BELIZE' 'ST. MAARTIN' 'GRAND CAYMAN' 'VANUATU' 'SRI LANKA'
 'URUGUAY' 'VENEZUELA' 'INDIA' 'OKINAWA' 'TANZANIA' ' TONGA'
 'MARS

Trimamos las strings para eliminar los espacios y convertimos a mayusculas para homogeneizar los registros

In [37]:
# Convert all values in the 'Country' column to uppercase and remove leading/trailing whitespace
df_clear['Country'] = df_clear['Country'].str.upper().str.strip()

# Display unique values to confirm the changes
valores_unicos = df_clear['Country'].unique()
print(valores_unicos)


['AUSTRALIA' 'USA' 'UNITED KINGDOM' 'BAHAMAS' 'UNITED ARAB EMIRATES'
 'MEXICO' 'SOUTH AFRICA' 'REUNION' 'ST HELENA, BRITISH OVERSEAS TERRITORY'
 'INDONESIA' 'ENGLAND' 'NEW CALEDONIA' 'SPAIN' 'CHINA' 'JAPAN' 'COLUMBIA'
 'EGYPT' 'NEW ZEALAND' 'FRENCH POLYNESIA' 'CAPE VERDE' 'FIJI' 'BRAZIL'
 'CAYMAN ISLANDS' 'ARUBA' 'ECUADOR' 'MOZAMBIQUE' 'THAILAND' 'PUERTO RICO'
 'ITALY' 'MAURITIUS' 'ST. MARTIN' 'PAPUA NEW GUINEA' 'KIRIBATI' 'ISRAEL'
 'FRANCE' 'DIEGO GARCIA' 'JAMAICA' 'PALESTINIAN TERRITORIES' 'GUAM'
 'SEYCHELLES' 'NIGERIA' 'TONGA' 'SCOTLAND' 'TRINIDAD & TOBAGO' 'CANADA'
 'SAUDI ARABIA' 'TAIWAN' 'DOMINICAN REPUBLIC' 'KENYA' 'RUSSIA'
 'PHILIPPINES' 'TURKS & CAICOS' 'COSTA RICA' 'UNITED ARAB EMIRATES (UAE)'
 'SAMOA' 'SOLOMON ISLANDS' 'SOUTH KOREA' 'MALTA' 'VIETNAM' 'MADAGASCAR'
 'PANAMA' 'AZORES' 'CUBA' 'CROATIA' 'BRITISH VIRGIN ISLANDS' 'NORWAY'
 'BELIZE' 'ST. MAARTIN' 'GRAND CAYMAN' 'VANUATU' 'SRI LANKA' 'URUGUAY'
 'VENEZUELA' 'INDIA' 'OKINAWA' 'TANZANIA' 'MARSHALL ISLANDS' 'SENEGAL'
 'H

In [38]:
df_clear.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3191 entries, 0 to 5953
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Year         3191 non-null   int32   
 1   Type         3191 non-null   category
 2   Country      3191 non-null   object  
 3   Sex          3191 non-null   object  
 4   Age          3191 non-null   int64   
 5   Fatal (Y/N)  3191 non-null   object  
 6   Date         3191 non-null   object  
 7   Activity     3191 non-null   object  
 8   Specie       3191 non-null   object  
 9   Month        3191 non-null   float64 
dtypes: category(1), float64(1), int32(1), int64(1), object(6)
memory usage: 240.2+ KB


Seguidamente trabajamos la columna Sex

se modifica el nombre de la columna porque contenia errores de sintaxis

In [39]:
df_clear["Sex"] = df_clear["Sex "]

df_clear = df_clear.drop("Sex ", axis=1)
df_clear.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3191 entries, 0 to 5953
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Year         3191 non-null   int32   
 1   Type         3191 non-null   category
 2   Country      3191 non-null   object  
 3   Age          3191 non-null   int64   
 4   Fatal (Y/N)  3191 non-null   object  
 5   Date         3191 non-null   object  
 6   Activity     3191 non-null   object  
 7   Specie       3191 non-null   object  
 8   Month        3191 non-null   float64 
 9   Sex          3191 non-null   object  
dtypes: category(1), float64(1), int32(1), int64(1), object(6)
memory usage: 240.2+ KB


In [40]:
valores_unicos = df_clear["Sex"].unique()
print(valores_unicos)

['M' 'F' 'M ' 'lli']


detectamos que hay mas tipos de generos de los esperados.

procedemos a reparar el genero masculino que presentra dos estados "M" y "M " 

In [41]:
df_clear.loc[df_clear["Sex"]=="M ", "Sex"]= "M"
valores_unicos = df_clear["Sex"].unique()
print(valores_unicos)

['M' 'F' 'lli']


In [42]:
cuantos_lli =(df_clear["Sex"] == "lli").sum()
print(cuantos_lli)

1


Como solo hay un lli vamos a ver si es Hombre o mujer 

In [43]:
mis_lli = df[df["Sex "]=="lli"]
mis_lli

Case Number        Date    Year        Type Country        Area  \
1484  2004.11.11.b  2004-11-11  2004.0  Unprovoked     USA  California   

                                            Location Activity        Name  \
1484  Bunkers, Humboldt Bay, Eureka, Humboldt County  Surfing  Brian Kang   

     Sex   ... Fatal (Y/N)   Time                 Species   \
1484  lli  ...           N  13h30  5.5 m [18'] white shark   

                Investigator or Source                    pdf  \
1484  R. Collier, GSAF                  2004.11.11.b-Kang.pdf   

                                           href formula  \
1484  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
1484  http://sharkattackfile.net/spreadsheets/pdf_di...  2004.11.11.b   

     Case Number.2 original order  
1484  2004.11.11.b         4611.0  

[1 rows x 22 columns]

como podemos observar es un hombre por lo tanto se le asignará el valor "M"

In [44]:
df_clear.loc[df_clear["Sex"]=="lli", "Sex"]= "M"

In [45]:
valores_unicos = df_clear["Sex"].unique()
print(valores_unicos)

['M' 'F']


Cambiamos el tipo de dato de la columna Sex de objeto a categoria

In [46]:
df_clear['Sex'] = df_clear['Sex'].astype('category')
df_clear.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3191 entries, 0 to 5953
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Year         3191 non-null   int32   
 1   Type         3191 non-null   category
 2   Country      3191 non-null   object  
 3   Age          3191 non-null   int64   
 4   Fatal (Y/N)  3191 non-null   object  
 5   Date         3191 non-null   object  
 6   Activity     3191 non-null   object  
 7   Specie       3191 non-null   object  
 8   Month        3191 non-null   float64 
 9   Sex          3191 non-null   category
dtypes: category(2), float64(1), int32(1), int64(1), object(5)
memory usage: 218.5+ KB


In [47]:
valores_unicos = df_clear["Fatal (Y/N)"].unique()
print(valores_unicos)

['N' 'Y' '2017' 'UNKNOWN' ' N']


cambiamos los valores de " N" a "N"

In [48]:
df_clear.loc[df_clear["Fatal (Y/N)"]==" N", "Fatal (Y/N)"]= "N"
valores_unicos = df_clear["Fatal (Y/N)"].unique()
print(valores_unicos)

['N' 'Y' '2017' 'UNKNOWN']


In [49]:
cuantos_2017 =(df_clear["Fatal (Y/N)"] == "2017").sum()
print(cuantos_2017)

cuantos_uknown =(df_clear["Fatal (Y/N)"] == "UNKNOWN").sum()
print(cuantos_uknown)

1
9


dada la marginalidad de los datos de Unknown y 2017 decidimos elimiar dichos registros

In [50]:
indiceseliminar = df_clear[(df_clear["Fatal (Y/N)"] == "2017") | (df_clear["Fatal (Y/N)"] == "UNKNOWN")].index

df_clear = df_clear.drop(indiceseliminar)



In [51]:
valores_unicos = df_clear["Fatal (Y/N)"].unique()
print(valores_unicos)

['N' 'Y']


Vamos a convertir la columna Fatal de objeto a Bool

In [52]:
df_clear['Fatal (Y/N)'] = df_clear['Fatal (Y/N)'].replace({"Y": True, "N": False})

df_clear.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3181 entries, 0 to 5953
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Year         3181 non-null   int32   
 1   Type         3181 non-null   category
 2   Country      3181 non-null   object  
 3   Age          3181 non-null   int64   
 4   Fatal (Y/N)  3181 non-null   bool    
 5   Date         3181 non-null   object  
 6   Activity     3181 non-null   object  
 7   Specie       3181 non-null   object  
 8   Month        3181 non-null   float64 
 9   Sex          3181 non-null   category
dtypes: bool(1), category(2), float64(1), int32(1), int64(1), object(4)
memory usage: 196.0+ KB


/tmp/ipykernel_45368/1774482131.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_clear['Fatal (Y/N)'] = df_clear['Fatal (Y/N)'].replace({"Y": True, "N": False})


Ahora vamos a comprobar si existen valores duplicados

In [53]:
duplicates = df_clear.duplicated().sum()
print(f"\nNumber of duplicate rows: {duplicates}")


Number of duplicate rows: 5


eliminamos los duplicados y obtenemos nuestro dataframe final listo para el analisis

In [54]:

df_finish = df_clear.drop_duplicates()

print("\nDuplicates Removed.")
df_finish.info()



Duplicates Removed.
<class 'pandas.core.frame.DataFrame'>
Index: 3176 entries, 0 to 5953
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Year         3176 non-null   int32   
 1   Type         3176 non-null   category
 2   Country      3176 non-null   object  
 3   Age          3176 non-null   int64   
 4   Fatal (Y/N)  3176 non-null   bool    
 5   Date         3176 non-null   object  
 6   Activity     3176 non-null   object  
 7   Specie       3176 non-null   object  
 8   Month        3176 non-null   float64 
 9   Sex          3176 non-null   category
dtypes: bool(1), category(2), float64(1), int32(1), int64(1), object(4)
memory usage: 195.7+ KB


Ultimamos los tipos de datos de algunas columnas

In [55]:
df_finish['Specie'] = df_finish['Specie'].astype('category')

/tmp/ipykernel_45368/3037862725.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_finish['Specie'] = df_finish['Specie'].astype('category')


In [56]:
df_finish['Activity'] = df_finish['Activity'].astype('category')

/tmp/ipykernel_45368/2431292850.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_finish['Activity'] = df_finish['Activity'].astype('category')


In [57]:
df_finish['Month'] = df_finish['Month'].astype('int32')
df_finish.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3176 entries, 0 to 5953
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Year         3176 non-null   int32   
 1   Type         3176 non-null   category
 2   Country      3176 non-null   object  
 3   Age          3176 non-null   int64   
 4   Fatal (Y/N)  3176 non-null   bool    
 5   Date         3176 non-null   object  
 6   Activity     3176 non-null   category
 7   Specie       3176 non-null   category
 8   Month        3176 non-null   int32   
 9   Sex          3176 non-null   category
dtypes: bool(1), category(4), int32(2), int64(1), object(2)
memory usage: 140.8+ KB


/tmp/ipykernel_45368/1257157135.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_finish['Month'] = df_finish['Month'].astype('int32')


Eliminamos la columna Date

In [58]:
df_finish = df_finish.drop("Date", axis=1)
df_finish.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3176 entries, 0 to 5953
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Year         3176 non-null   int32   
 1   Type         3176 non-null   category
 2   Country      3176 non-null   object  
 3   Age          3176 non-null   int64   
 4   Fatal (Y/N)  3176 non-null   bool    
 5   Activity     3176 non-null   category
 6   Specie       3176 non-null   category
 7   Month        3176 non-null   int32   
 8   Sex          3176 non-null   category
dtypes: bool(1), category(4), int32(2), int64(1), object(1)
memory usage: 115.9+ KB


In [59]:
# Save with a custom delimiter and exclude index
df_finish.to_csv('Shark_finish.csv', index=False, header=True)
